### PR TITLE
Exclude pinned tabs when closing items

### LINKED ISSUE #19560: Close all except pinned tabs
### Check for existing issues

- [x] Completed

### Describe the feature

Have `:qa` and `pane:close all items` avoid closing pinned tabs

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by axelcarl**
Closing multiple items will no longer closed pinned tabs.

Closes #19560

Release Notes:

- Fixed close actions closing pinned tabs.

### CODE REVIEW

**Comment by CharlesChen0823** [comment]
might we can add one more action? like `CloseAllItemsExcludePinned`?

**Comment by axelcarl** [comment]
@CharlesChen0823 that could be an option, or add an action called CloseAllIncludingPinned, I guess it depends on what is the most sensible default?

**Comment by HarshNarayanJha** [comment]
Well, Close All Tabs shouldn't really close the Pinned tabs because they are _special_, that's why they were pinned. Adding an option that says `Close All Except Pinned` will only introduce inconvenience!

**Comment by axelcarl** [comment]
Good point, will add the action! :)

**Comment by HarshNarayanJha** [comment]
> Do you think I should add a CloseAllIncludingPinned that can be bound to qa (vim-mode) for those that want that as their default behavior?

Could be useful for is non vim users too. I agree with CloseAllIncludingPinned Action. But don't make it default, and for vim users, they know it better.

**Comment by axelcarl** [comment]
@CharlesChen0823 @HarshNarayanJha Added the new action "CloseAllItemsIncludingPinned" and tests, I guess the question now is how close left/right should be affected, it would maybe make sense to alter them in a separate pull request though.

**Review by SomeoneToIgnore — COMMENTED** [review]
1. I think we need to adjust all "Close*` tab-related commands
2. We do not need any extra actions (not sure why that was suggested initially, but it's not a maintainer's suggestion so feel free to ignore).
Something like
```
#[derive(Clone, PartialEq, Debug, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CloseAllItems {
    pub save_intent: Option<SaveIntent>,
    pub close_pinned: bool,
}
```

where `close_pinned` will be `false` by default and anyone may rebind it to something else if they need.

**Comment by axelcarl** [comment]
@SomeoneToIgnore Good idea, will refactor the PR.

File: crates/workspace/src/pane.rs [hunk 1]
```diff
@@ -105,12 +105,32 @@ pub struct CloseActiveItem {
...
 pub struct CloseInactiveItems {
     pub save_intent: Option<SaveIntent>,
+    pub close_pinned: Option<bool>,
 }
 
 #[derive(Clone, PartialEq, Debug, Deserialize, Default)]
 #[serde(rename_all = "camelCase")]
 pub struct CloseAllItems {
     pub save_intent: Option<SaveIntent>,
+    pub close_pinned: Option<bool>,
```
Comment (SomeoneToIgnore) [inline crates/workspace/src/pane.rs:115]:
```suggestion
    pub close_pinned: bool,
```

Why cannot it be just this? What does `None` mean?

Same for everything below.

**Review by SomeoneToIgnore — COMMENTED** [review]
Nice, and even the new test is there.
Seems to be close to merging.

Let's also update both `default-macos.json` and `default-linux.json` to explicitly show the `"close_pinned": false` in each affected action: this will serve as a documentation.

File: crates/workspace/src/pane.rs (hunk 1 above)
Comment (axelcarl) [reply crates/workspace/src/pane.rs:115]: I opted for an Option to keep support for old keymaps, however if that is not reason enough I can switch it to bool!

(Setting it to a bool would require that every CloseAction would have to be followed by brackets and an explicit boolean value, whereas leaving it empty now would mean close_pinned would be "false" due to the unwrap_or statements).

Comment (SomeoneToIgnore) [reply crates/workspace/src/pane.rs:115]: > Setting it to a bool would require that every CloseAction would have to be followed by brackets and an explicit boolean value

Can you elaborate? 
I think it's not true at all, as when we deserialize, we fill in the defaults for the missing declarations.
Example:
we have https://github.com/zed-industries/zed/blob/b0b29d91f9a21870dc48932ae54ad5f812298bd4/assets/keymaps/default-macos.json#L117

and when I declare
```
{
    "bindings": {
      "shift-d": "editor::SelectToBeginningOfLine"
    }
  }
```

it works for me juts fine.

Comment (axelcarl) [reply crates/workspace/src/pane.rs:115]: Alright, then I was mistaken, will alter it to bool and update the json files. Thanks!

**Review by SomeoneToIgnore — APPROVED** [review]
Thank you.

**Pull request merged**